test/src/e2e_vm_tests/test_programs/should_fail/expression_cannot_be_assigned_to/test.toml

```diff
@@ -0,0 +1,75 @@
+category = "fail"
+
+#check: $()Expression cannot be assigned to
+#check: $()S { x: 0 } = 0;
+#nextln: $()This expression cannot be assigned to, because it is a struct instantiation.
+#check: $()Struct instantiation cannot be an assignment target.
+
+#check: $()Expression cannot be assigned to
+#check: $()S { x: 0 }.x = 0;
+#nextln: $()This expression cannot be assigned to, because it contains a struct instantiation.
+#nextln: $()This is the contained struct instantiation.
+#check: $()Struct instantiation cannot be a part of an assignment target.
+
+#check: $()Expression cannot be assigned to
+#check: $()return_array() = 1;
+#nextln: $()This expression cannot be assigned to, because it is a function call.
+#check: $()Function call cannot be an assignment target.
+
+#check: $()Expression cannot be assigned to
```

Comment:
Is this a temporary thing? I can see this being useful for Vecs and Maps:

```sway
map.get(key).value += 1;
```

Reply:
Currently, without dereferencing our LHSs were very restricted. Compared to Rust at least. So on one side, yes we can say it's just not implemented.

On the other side, supporting arbitrary expressions in element access LHS (means no dereferencing, just expressions that in the end provide indices to the value that needs to be assigned to) can lead to more confusion than actually bring any benefits.

The reason for this is, that in a case of an arbitrary expression like e.g., `S { x: 0 }.x = 1` the assignment will change a temporary value.

E.g., in the above example of `map.get(key)`, I guess the interface for `Map` would be similar with the one for `Vec` and `get()` would return `Option<T>`. But let's assume it returns `T` or panics otherwise. The following expression, if allowed, looks like it will change the value behind the `key`:
```
map.get(key) += 1;
```
But it will expand to:
```
map.get(key) = map.get(key) + 1;
```
where both `get`s above return a _copy_ of the value contained in the `map`. thus making no actual change to the value stored in the `map`.

In case of collections, I expect them to start using references once they are fully supported, and provide `get()` methods that return references to contained values.

That way, the above code would be:
```
*map.get(key) += 1;
```
and it would change the stored value as expected.